### PR TITLE
SI-3871 Tests for protected access

### PR DESCRIPTION
**Pull request by retronym**
c39f26382dddaa7 fixed the bug but didn't commit a test case.

While I was looking at this, I added a test to exercise protected more thoroughly.

Review by @gkossakowski

### CODE REVIEW

File: test/files/neg/t3781b.scala [hunk 1]
```diff
@@ -0,0 +1,121 @@
+/**
```
Comment (som-snytt) [inline test/files/neg/t3781b.scala]: Transpose typo in test file name, `t3871`

Comment (retronym) [inline test/files/neg/t3781b.scala]: Thanks, fixed.

**Comment by retronym** [comment]
PLS REBUILD ALL

File: test/files/neg/t3871b.scala [hunk 2]
```diff
@@ -0,0 +1,121 @@
+/**
+A protected identifier x may be used as a member name in a selection r.x only
+if one of the following applies:
+  1a. The access is within the template defining the member, or,
+   b. if a qualification C is given, inside the package C, or
+   c. the class C , or its companion module, or
```
Comment (gkossakowski) [inline test/files/neg/t3871b.scala]: Not sure I understand that statement. What does "c. the class C , or its companion module, or" mean? Can it be interpreted alone?

Comment (gkossakowski) [inline test/files/neg/t3871b.scala]: I just checked the spec and now it makes sense to me. I'd suggest rewording your comment to be:

```
1a. The access is within the template defining the member, or,
if a qualification C is given,
1b. inside the package C, or
1c. the class C , or its companion module, or
```

or something along those lines. The point is that "if qualification is given" applies both to 1b and 1c.

File: test/files/neg/t3871b.scala [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    def this(a: Any) = {this(); prot; protE; protT}
+    object B extends A {
+      A.this.prot
+      A.this.protE
+      A.this.protT
+    }
+
+    import A._
+    // !!! Why is this allowed? 1? 3?
+    protO
```
Comment (gkossakowski) [inline test/files/neg/t3871b.scala]: I believe this is due to the following fragment of the spec (not quoted in your comment above):

> The protected modifier applies to class member definitions. Protected members of a class can be accessed from within
> - the companion module of any of those classes

File: test/files/neg/t3871b.check [hunk 4]
```diff
@@ -0,0 +1,97 @@
+t3871b.scala:55: error: not found: value protOT
```
Comment (gkossakowski) [inline test/files/neg/t3871b.check]: Should it say not found or complain about access rules as in other cases?

File: test/files/neg/t3871b.scala [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+      a.protT   // not allowed
+    }
+  }
+  object B {
+    def n(a: A, b: B, c: C) = {
+      b.prot    // 3 !!!
+      c.prot    // 3 !!!
+      // Wording of 3 seems insufficient, missing:
+      // "... (if the access is from a class), or
+      // the type instance of companion class (if the access is from a module)"
```
Comment (gkossakowski) [inline test/files/neg/t3871b.scala:90]: Agreed.

**Comment by gkossakowski** [comment]
@retronym: impressive test case! :+1: 

I reviewed everything and I got only minor comments that you can see inline.

File: test/files/neg/t3871b.scala (hunk 2 above)
Comment (som-snytt) [inline test/files/neg/t3871b.scala]: In the spec, the previous two spots with that language is clearer: "...accessible respectively from ... package C or ... class C".

This is pretty interesting.  Maybe too interesting.

File: test/files/neg/t3871b.scala (hunk 3 above)
Comment (retronym) [inline test/files/neg/t3871b.scala]: Right you are.

**Comment by gkossakowski** [comment]
@retronym: ping

**Comment by gkossakowski** [comment]
@retronym: ping

**Comment by retronym** [comment]
Sorry about the delay. I've addressed the first two comments. I don't know why one of the error messages doesn't give the hint about access. I agree it seems inconsistent.
